### PR TITLE
Improve checks for property setters in Python API

### PR DESCRIPTION
**Pull request by paulromano**
This pull request improves type checks on setter methods for many class properties. There are a few aspects to this:

- Attributes that were array-like were mostly checked to see if they were either a tuple, list, or ndarray. The basic functionality that we want to ensure is that whatever is being passed is iterable, so now these attributes are checked against `collections.Iterable`. Originally, I had used `collections.Sequence` which seems more appropriate, but unfortunately ndarrays do not implement all required abstract methods of Sequence.
- Strings were being checked against str and numpy.str (same as str actually, just try `np.str is str`). This actually excludes unicode strings in Python 2. Now strings are checked against basestring.
- The same logic for type and length checks appeared all over the API. I've modularized it in openmc.checkvalue with check_type, check_length, and check_value functions. This reduces about 1000 lines of code.

### CODE REVIEW

File: openmc/cmfd.py [hunk 1]
```diff
@@ -417,175 +337,90 @@ def write_matrices(self):
...
         self._display = display
 
     @downscatter.setter
     def downscatter(self, downscatter):
-        if not isinstance(downscatter, bool):
-            msg = 'Unable to set downscatter to {0} which is ' \
-                  'a non-boolean value'.format(downscatter)
-            raise ValueError(msg)
-
+        check_type('downscatter', downscatter, bool)
```
Comment (wbinventor) [inline openmc/cmfd.py]: Perhaps prefix the ``name`` param with "CMFD" here as you do elsewhere?

File: openmc/cmfd.py [hunk 2]
```diff
@@ -417,175 +337,90 @@ def write_matrices(self):
...
         self._begin = begin
 
     @dhat_reset.setter
     def dhat_reset(self, dhat_reset):
-        if not isinstance(dhat_reset, bool):
-            msg = 'Unable to set Dhat reset to {0} which is ' \
-                  'a non-boolean value'.format(dhat_reset)
-            raise ValueError(msg)
-
+        check_type('Dhat reset', dhat_reset, bool)
```
Comment (wbinventor) [inline openmc/cmfd.py]: Perhaps prefix the ``name`` param with "CMFD" here as you do elsewhere?

File: openmc/cmfd.py [hunk 3]
```diff
@@ -417,175 +337,90 @@ def write_matrices(self):
...
         self._gauss_seidel_tolerance = gauss_seidel_tolerance
 
     @ktol.setter
     def ktol(self, ktol):
-        if not is_integer(ktol) and not is_float(ktol):
-            msg = 'Unable to set the eigenvalue tolerance to {0} which is ' \
-                  'not an integer or floating point value'.format(ktol)
-            raise ValueError(msg)
-
+        check_type('eigenvalue tolerance', ktol, Real)
```
Comment (wbinventor) [inline openmc/cmfd.py]: ditto "CMFD" prefix

File: openmc/cmfd.py [hunk 4]
```diff
@@ -417,175 +337,90 @@ def write_matrices(self):
...
-                  'it must be of length 2'.format(width)
-            raise ValueError(msg)
-
-        for t in gauss_seidel_tolerance:
-            if not is_integer(t) and not is_float(t):
-                msg = 'Unable to set Gauss-Seidel tolerance with {0} which ' \
-                      'is not an integer or floating point value'.format(t)
-                raise ValueError(msg)
-
+        check_type('Gauss-Seidel tolerance', gauss_seidel_tolerance,
```
Comment (wbinventor) [inline openmc/cmfd.py]: ditto "CMFD" prefix

File: openmc/element.py [hunk 5]
```diff
@@ -59,16 +64,12 @@ def name(self):
...
-                  'with a non-string value {0}'.format(xs)
-            raise ValueError(msg)
-
+        check_type('cross section identifier', xs, basestring)
         self._xs = xs
 
     @name.setter
     def name(self, name):
-        if not is_string(name):
+        if not isinstance(name, basestring):
```
Comment (wbinventor) [inline openmc/element.py]: Why can't you use ``check_type`` for the ``name``?

File: openmc/filter.py [hunk 6]
```diff
@@ -132,18 +135,18 @@ def bins(self, bins):
...
                 elif edge < 0:
                     msg = 'Unable to add bin "{0}" to a {1} Filter since ' \
-                          'it is a negative integer'.format(edge, self._type)
+                          'it is negative'.format(edge, self._type)
                     raise ValueError(msg)
 
         elif self._type in ['energy', 'energyout']:
             for edge in bins:
-                if not is_integer(edge) and not is_float(edge):
+                if not isinstance(edge, Real):
```
Comment (wbinventor) [inline openmc/filter.py]: Cannot ``check_type`` be used for the bin edges?

File: openmc/filter.py [hunk 7]
```diff
@@ -132,18 +135,18 @@ def bins(self, bins):
         if self._type in ['cell', 'cellborn', 'surface', 'material',
                           'universe', 'distribcell']:
             for edge in bins:
-                if not is_integer(edge):
+                if not isinstance(edge, Integral):
```
Comment (wbinventor) [inline openmc/filter.py]: Ditto cannot ``check_type`` be used for the bin edges here?

File: openmc/filter.py [hunk 8]
```diff
@@ -132,18 +135,18 @@ def bins(self, bins):
...
                     msg = 'Unable to add bin "{0}" to a {1} Filter since ' \
-                          'it is a non-integer'.format(edge, self._type)
+                          'it is not an integer'.format(edge, self._type)
                     raise ValueError(msg)
                 elif edge < 0:
                     msg = 'Unable to add bin "{0}" to a {1} Filter since ' \
-                          'it is a negative integer'.format(edge, self._type)
+                          'it is negative'.format(edge, self._type)
```
Comment (wbinventor) [inline openmc/filter.py]: It might be useful to have a ``check_less_than(...)`` and a ``check_greater_than(...)`` as well

File: openmc/filter.py [hunk 9]
```diff
@@ -203,20 +203,13 @@ def mesh(self, mesh):
 
     @offset.setter
     def offset(self, offset):
-        if not is_integer(offset):
-            msg = 'Unable to set offset "{0}" for a {1} Filter since it is a ' \
-                  'non-integer value'.format(offset, self._type)
-            raise ValueError(msg)
+        check_type('offset', offset, Integral)
```
Comment (wbinventor) [inline openmc/filter.py]: append "filter" to the ``name``

File: openmc/filter.py [hunk 10]
```diff
@@ -192,10 +195,7 @@ def num_bins(self, num_bins):
 
     @mesh.setter
     def mesh(self, mesh):
-        if not isinstance(mesh, Mesh):
-            msg = 'Unable to set Mesh to "{0}" for Filter since it is not a ' \
-                  'Mesh object'.format(mesh)
-            raise ValueError(msg)
+        check_type('mesh', mesh, Mesh)
```
Comment (wbinventor) [inline openmc/filter.py]: append "filter" to the ``name``

File: openmc/filter.py [hunk 11]
```diff
@@ -203,20 +203,13 @@ def mesh(self, mesh):
...
         self._offset = offset
 
     @stride.setter
     def stride(self, stride):
-        if not is_integer(stride):
-            msg = 'Unable to set stride "{0}" for a {1} Filter since it is a ' \
-                  'non-integer value'.format(stride, self._type)
-            raise ValueError(msg)
-
+        check_type('stride', stride, Integral)
```
Comment (wbinventor) [inline openmc/filter.py]: append "filter" to the ``name``

File: openmc/filter.py (hunk 6 above)
Comment (paulromano) [inline openmc/filter.py]: It could, but the default error message wouldn't make sense, i.e. "can't set __ to ___".

File: openmc/material.py [hunk 12]
```diff
@@ -115,35 +119,25 @@ def id(self, material_id):
...
-                  'non-negative integer'.format(material_id)
-            raise ValueError(msg)
-
         else:
+            check_type('material ID', material_id, Integral)
+            if material_id in MATERIAL_IDS:
+                msg = 'Unable to set Material ID to {0} since a Material with ' \
+                      'this ID was already initialized'.format(material_id)
+                raise ValueError(msg)
+            elif material_id < 0:
```
Comment (wbinventor) [inline openmc/material.py]: ditto ``check_less_than(...)``

File: openmc/mesh.py [hunk 13]
```diff
@@ -31,15 +36,15 @@ class Mesh(object):
         Name of the mesh
     type : str
         Type of the mesh
-    dimension : tuple or list or ndarray
+    dimension : Sequnce of int
```
Comment (wbinventor) [inline openmc/mesh.py]: typo: "Sequnce" -> "Sequence" (as well as several more below)

File: openmc/mesh.py [hunk 14]
```diff
@@ -135,128 +140,49 @@ def id(self, mesh_id):
...
-            raise ValueError(msg)
-
-        elif mesh_id < 0:
-            msg = 'Unable to set Mesh ID to {0} since it must be a ' \
-                  'non-negative integer'.format(mesh_id)
-            raise ValueError(msg)
-
         else:
+            check_type('mesh ID', mesh_id, Integral)
+            if mesh_id < 0:
```
Comment (wbinventor) [inline openmc/mesh.py]: ditto ``check_less_than(...)``

File: openmc/plots.py [hunk 15]
```diff
@@ -136,174 +141,74 @@ def id(self, plot_id):
...
-            raise ValueError(msg)
-
-        elif plot_id < 0:
-            msg = 'Unable to set Plot ID to {0} since it must be a ' \
-                  'non-negative integer'.format(plot_id)
-            raise ValueError(msg)
-
         else:
+            check_type('plot ID', plot_id, Integral)
+            if plot_id < 0:
```
Comment (wbinventor) [inline openmc/plots.py]: ditto ``check_less_than(...)``

File: openmc/plots.py [hunk 16]
```diff
@@ -136,174 +141,74 @@ def id(self, plot_id):
...
+        check_type('plot pixels', pixels, Iterable, Integral)
+        check_length('plot pixels', pixels, 2, 3)
         for dim in pixels:
-            if not is_integer(dim):
-                msg = 'Unable to create Plot ID={0} with pixel value {1} ' \
-                      'which is not an integer'.format(self._id, dim)
-                raise ValueError(msg)
-
-            elif dim < 0:
+            if dim < 0:
```
Comment (wbinventor) [inline openmc/plots.py]: ditto ``check_less_than(...)``

File: openmc/plots.py [hunk 17]
```diff
@@ -136,174 +141,74 @@ def id(self, plot_id):
...
+        check_type('plot background', background, Iterable, Integral)
+        check_length('plot background', background, 3)
         for rgb in background:
-            if not is_integer(rgb):
-                msg = 'Unable to create Plot ID={0} with background RGB ' \
-                      'value {1} which is not an integer'.format(self._id, rgb)
-                raise ValueError(msg)
-
-            elif rgb < 0 or rgb > 255:
+            if rgb < 0 or rgb > 255:
```
Comment (wbinventor) [inline openmc/plots.py]: ditto ``check_less_than(...)`` and ``check_greater_than(...)``

File: openmc/plots.py [hunk 18]
```diff
@@ -341,52 +245,24 @@ def col_spec(self, col_spec):
...
+        check_length('plot mask background', mask_background, 3)
         for rgb in mask_background:
-
-            if not is_integer(rgb):
-                msg = 'Unable to create Plot ID={0} with mask background RGB ' \
-                      'value {1} which is not an integer'.format(self._id, rgb)
-                raise ValueError(msg)
-
-            elif rgb < 0 or rgb > 255:
+            if rgb < 0 or rgb > 255:
```
Comment (wbinventor) [inline openmc/plots.py]: ditto ``check_less_than(...)`` and ``check_greater_than(...)``

File: openmc/settings.py [hunk 19]
```diff
@@ -395,58 +400,38 @@ def dd_count_interactions(self):
...
         self._batches = batches
 
     @generations_per_batch.setter
     def generations_per_batch(self, generations_per_batch):
-        if not is_integer(generations_per_batch):
-            msg = 'Unable to set generations per batch to a non-integer ' \
-                  'value {0}'.format(generations_per_batch)
-            raise ValueError(msg)
-
+        check_type('generation per patch', generations_per_batch, Integral)
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 20]
```diff
@@ -512,37 +493,15 @@ def set_source_space(self, stype, params):
...
-
-        for param in params:
-            if not is_integer(param) and not is_float(param):
-                msg = 'Unable to set source space parameters to {0} since it ' \
-                      'is not an integer or floating point value'.format(param)
-                raise ValueError(msg)
+        check_type('source space type', stype, basestring)
+        check_value('source space type', stype, ['box', 'fission', 'point'])
+        check_type('source space parameters', params, Iterable, Real)
+        if stype in ['box', 'fission'] and len(params) != 6:
```
Comment (wbinventor) [inline openmc/settings.py]: why does this conditional include ``len(params) != 6`` only to then call ``check_length(...)``?

File: openmc/settings.py [hunk 21]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
 
     @verbosity.setter
     def verbosity(self, verbosity):
-        if not is_integer(verbosity):
-            msg = 'Unable to set verbosity to non-integer ' \
-                  'value {0}'.format(verbosity)
-            raise ValueError(msg)
-
+        check_type('verbosity', verbosity, Integral)
         if verbosity < 1 or verbosity > 10:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)`` and ``check_greater_than(...)``

File: openmc/settings.py [hunk 22]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
 
     @seed.setter
     def seed(self, seed):
-        if not is_integer(seed):
-            msg = 'Unable to set seed to non-integer value {0}'.format(seed)
-            raise ValueError(msg)
-
-        elif seed <= 0:
+        check_type('random number generator seed', seed, Integral)
+        if seed <= 0:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``?

File: openmc/settings.py [hunk 23]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
     @weight.setter
     def weight(self, weight):
-        if not is_float(weight):
-            msg = 'Unable to set weight cutoff to non-floating point ' \
-                  'value {0}'.format(weight)
-            raise ValueError(msg)
-
-        elif weight < 0.0:
+        check_type('weight cutoff', weight, Real)
+        if weight < 0.0:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 24]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
 
     @weight_avg.setter
     def weight_avg(self, weight_avg):
-        if not is_float(weight_avg):
-            msg = 'Unable to set weight avg. to non-floating point ' \
-                  'value {0}'.format(weight_avg)
-            raise ValueError(msg)
-        elif weight_avg < 0.0:
+        check_type('average survival weight', weight_avg, Real)
+        if weight_avg < 0.0:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 25]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
     @trigger_max_batches.setter
     def trigger_max_batches(self, trigger_max_batches):
-        if not is_integer(trigger_max_batches):
-            msg = 'Unable to set trigger max batches to a non-integer ' \
-                  'value {0}'.format(trigger_max_batches)
-            raise ValueError(msg)
-
-        elif trigger_max_batches <= 0:
+        check_type('trigger maximum batches', trigger_max_batches, Integral)
+        if trigger_max_batches <= 0:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 26]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
     @trigger_batch_interval.setter
     def trigger_batch_interval(self, trigger_batch_interval):
-        if not is_integer(trigger_batch_interval):
-            msg = 'Unable to set trigger batch interval to a non-integer ' \
-                  'value {0}'.format(trigger_batch_interval)
-            raise ValueError(msg)
-
-        elif trigger_batch_interval <= 0:
+        check_type('trigger batch interval', trigger_batch_interval, Integral)
+        if trigger_batch_interval <= 0:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 27]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
     @threads.setter
     def threads(self, threads):
-        if not is_integer(threads):
-            msg = 'Unable to set the threads to a non-integer ' \
-                  'value {0}'.format(threads)
-            raise ValueError(msg)
-
-        elif threads <= 0:
+        check_type('number of threads', threads, Integral)
+        if threads <= 0:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 28]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
-
-        elif len(trace) != 3:
-            msg = 'Unable to set the trace to {0} since it does not contain ' \
-                  '3 elements - batch, generation, and particle'.format(trace)
-            raise ValueError(msg)
-
-        elif trace[0] < 1:
+        check_type('trace', trace, Iterable, Integral)
+        check_length('trace', trace, 3)
+        if trace[0] < 1:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 29]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
-                  'be greater than or equal to 1'.format(track[2])
-            raise ValueError(msg)
-
+        check_type('track', track, Iterable, Integral)
+        if len(track) % 3 != 0:
+            msg = 'Unable to set the track to {0} since its length is ' \
+                  'not a multiple of 3'.format(track)
+            raise ValueError(msg)
+        for t in zip(track[::3], track[1::3], track[2::3]):
+            if t[0] < 1:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/settings.py [hunk 30]
```diff
@@ -694,433 +605,238 @@ def output(self, output):
...
-
+        check_type('UFS mesh dimension', dimension, Iterable, Integral)
+        check_length('UFS mesh dimension', dimension, 3)
         for dim in dimension:
-            if not is_integer(dim):
-                msg = 'Unable to set entropy mesh dimension to a ' \
-                      'non-integer {0}'.format(dim)
-                raise ValueError(msg)
-            elif dim < 1:
+            if dim < 1:
```
Comment (wbinventor) [inline openmc/settings.py]: ditto ``check_less_than(...)``

File: openmc/surface.py [hunk 31]
```diff
@@ -89,46 +93,24 @@ def id(self, surface_id):
...
-            raise ValueError(msg)
-
-        elif surface_id < 0:
-            msg = 'Unable to set Surface ID to {0} since it must be a ' \
-                  'non-negative integer'.format(surface_id)
-            raise ValueError(msg)
-
         else:
+            check_type('surface ID', surface_id, Integral)
+            if surface_id < 0:
```
Comment (wbinventor) [inline openmc/surface.py]: ditto ``check_less_than(...)``

File: openmc/surface.py [hunk 32]
```diff
@@ -600,20 +542,12 @@ def z0(self):
...
         self._coeffs['x0'] = x0
 
     @z0.setter
     def z0(self, z0):
-        if not is_integer(z0) and not is_float(z0):
-            msg = 'Unable to set z0 coefficient for YCylinder ID={0} to a ' \
-                  'non-integer value {1}'.format(self._id, z0)
-            raise ValueError(msg)
-
+        check_type('y0 coefficient', y0, Real)
```
Comment (wbinventor) [inline openmc/surface.py]: This is incorrect - it should be checking ``z0`` rather than ``y0``

File: openmc/tallies.py [hunk 33]
```diff
@@ -301,29 +303,18 @@ def id(self, tally_id):
...
-            raise ValueError(msg)
-
-        elif tally_id < 0:
-            msg = 'Unable to set Tally ID to {0} since it must be a ' \
-                  'non-negative integer'.format(tally_id)
-            raise ValueError(msg)
-
         else:
+            check_type('tally ID', tally_id, Integral)
+            if tally_id < 0:
```
Comment (wbinventor) [inline openmc/tallies.py]: ditto ``check_less_than(...)``

File: openmc/tallies.py [hunk 34]
```diff
@@ -383,45 +372,27 @@ def num_score_bins(self, num_score_bins):
...
     def num_realizations(self, num_realizations):
-        if not is_integer(num_realizations):
-            msg = 'Unable to set the number of realizations to "{0}" for ' \
-                  'Tally ID={1} since it is not an ' \
-                  'integer'.format(num_realizations)
-            raise ValueError(msg)
-
-        elif num_realizations < 0:
+        check_type('number of realizations', num_realizations, Integral)
+        if num_realizations < 0:
```
Comment (wbinventor) [inline openmc/tallies.py]: ditto ``check_less_than(...)``

File: openmc/tallies.py [hunk 35]
```diff
@@ -1338,7 +1309,7 @@ def export_results(self, filename='tally-results', directory='.',
                   '"{1}" since it is not supported'.format(self.id, format)
             raise ValueError(msg)
 
-        elif not isinstance(append, (bool, np.bool)):
+        elif not isinstance(append, bool):
```
Comment (wbinventor) [inline openmc/tallies.py]: Typo in my original error message here: "parameters" -> "parameter"

File: openmc/universe.py [hunk 36]
```diff
@@ -104,33 +110,22 @@ def id(self, cell_id):
...
-            raise ValueError(msg)
-
-        elif cell_id < 0:
-            msg = 'Unable to set Cell ID to {0} since it must be a ' \
-                        'non-negative integer'.format(cell_id)
-            raise ValueError(msg)
-
         else:
+            check_type('cell ID', cell_id, Integral)
+            if cell_id < 0:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 37]
```diff
@@ -476,30 +434,18 @@ def id(self, universe_id):
...
-            raise ValueError(msg)
-
-        elif universe_id < 0:
-            msg = 'Unable to set Universe ID to {0} since it must be a ' \
-                  'non-negative integer'.format(universe_id)
-            raise ValueError(msg)
-
         else:
+            check_type('universe ID', universe_id, Integral)
+            if universe_id < 0:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 38]
```diff
@@ -730,47 +676,27 @@ def id(self, lattice_id):
...
-            raise ValueError(msg)
-
-        elif lattice_id < 0:
-            msg = 'Unable to set Lattice ID to {0} since it must be a ' \
-                  'non-negative integer'.format(lattice_id)
-            raise ValueError(msg)
-
         else:
+            check_type('lattice ID', lattice_id, Integral)
+            if lattice_id < 0:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 39]
```diff
@@ -908,90 +834,35 @@ def offsets(self):
...
+        check_length('lattice dimension', dimension, 2, 3)
         for dim in dimension:
-            if not is_integer(dim) and not is_float(dim):
-                msg = 'Unable to set RectLattice ID={0} dimension to {1} since ' \
-                      'it is not an integer or floating point ' \
-                      'value'.format(self._id, dim)
-                raise ValueError(msg)
-
-            elif dim < 0:
+            if dim < 0:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 40]
```diff
@@ -908,90 +834,35 @@ def offsets(self):
...
 
     @offsets.setter
     def offsets(self, offsets):
-        if not isinstance(offsets, (tuple, list, np.ndarray)):
-            msg = 'Unable to set Lattice ID={0} offsets to {1} since ' \
-                  'it is not a Python tuple/list or NumPy ' \
-                  'array'.format(self._id, offsets)
-            raise ValueError(msg)
-
+        check_type('offsets', offsets, Iterable)
```
Comment (wbinventor) [inline openmc/universe.py]: clarify ``name`` param as "lattice offsets"

File: openmc/universe.py [hunk 41]
```diff
@@ -908,90 +834,35 @@ def offsets(self):
...
+        check_length('lattice pitch', pitch, 2, 3)
         for dim in pitch:
-            if not is_integer(dim) and not is_float(dim):
-                msg = 'Unable to set Lattice ID={0} pitch to {1} since ' \
-                      'it is not an an integer or floating point ' \
-                      'value'.format(self._id, dim)
-                raise ValueError(msg)
-
-            elif dim < 0:
+            if dim < 0:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 42]
```diff
@@ -1184,8 +1055,7 @@ def center(self):
 
     @num_rings.setter
     def num_rings(self, num_rings):
-
-        if not is_integer(num_rings) and num_rings < 1:
+        if not isinstance(num_rings, Integral) and num_rings < 1:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 43]
```diff
@@ -1194,7 +1064,7 @@ def num_rings(self, num_rings):
 
     @num_axial.setter
     def num_axial(self, num_axial):
-        if not is_integer(num_axial) and num_axial < 1:
+        if not isinstance(num_axial, Integral) and num_axial < 1:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

File: openmc/universe.py [hunk 44]
```diff
@@ -1203,52 +1073,19 @@ def num_axial(self, num_axial):
...
+        check_length('lattice pitch', pitch, 2, 3)
         for dim in pitch:
-            if not is_integer(dim) and not is_float(dim):
-                msg = 'Unable to set Lattice ID={0} pitch to {1} since ' \
-                      'it is not an an integer or floating point ' \
-                      'value'.format(self._id, dim)
-                raise ValueError(msg)
-
-            elif dim < 0:
+            if dim < 0:
```
Comment (wbinventor) [inline openmc/universe.py]: ditto ``check_less_than(...)``

**Comment by wbinventor** [comment]
Wow, you've very effectively gutted the entire Python API @paulromano! Nice job, I like how much this PR cleans up the code yet retains the ability to provide fine-grained user feedback.

By and large I think this is ready to go. I made a number of comments, but as you'll notice, most of them refer to the need for ``check_less_than(...)`` and ``check_greater_than(...)`` routines.

File: openmc/settings.py (hunk 20 above)
Comment (paulromano) [inline openmc/settings.py]: I was going to get rid of that... then I forgot.

**Comment by paulromano** [comment]
Ok, I think I've addressed everything. My convention for `check_greater_than` and `check_less_than` is the reverse of what you suggested. In my implementation, if you expect a value to be greater than zero, you'd call `check_greater_than`.

**Comment by wbinventor** [comment]
This looks good to me @paulromano! All of the examples run and the BEAVRS benchmarks still builds correctly.

**Pull request merged**